src/qunit.cpp

```diff
@@ -756,13 +756,118 @@ void QUnit::UniformlyControlledSingleBit(
     delete[] mappedControls;
 }
 
-void QUnit::X(bitLenInt qubit)
+void QUnit::X(bitLenInt target)
 {
-    shards[qubit].prob = ONE_R1 - shards[qubit].prob;
-    shards[qubit].unit->X(shards[qubit].mapped);
+    shards[target].prob = ONE_R1 - shards[target].prob;
+    shards[target].unit->X(shards[target].mapped);
 }
 
-void QUnit::Z(bitLenInt qubit) { shards[qubit].unit->Z(shards[qubit].mapped); }
+void QUnit::Z(bitLenInt target) { shards[target].unit->Z(shards[target].mapped); }
+
+void QUnit::CNOT(bitLenInt control, bitLenInt target)
+{
+    bitLenInt controls[1] = { control };
```

Comment:
Considering the very substantial amount of duplication in the next set of functions, can they be reduced down to a few utility functions with parameters for the variation?

Reply:
Notice that they already have, along with everything else that relies on `ApplyEitherControlled()`. It is difficult to make these implementations more terse than that. The performance gains absolutely justify the small amount of unavoidable repetition, in my opinion.